First cut at a 'limitations' section to the CHERI C/C++ guide [DRAFT]

rwatson opened the pull request:
This change introduces a new top-level 'Limitations to CHERI C/C++ memory safety' section to the CHERI C/C++ Programming Guide. The purpose of the new section are multifold but include:

- Documenting known gaps in CHERI memory safety for C/C++, which includes fundamental architectural, language, compiler, and runtime limitations -- such as around stack temporal safety, bounds precision, and compiler optimisation.
- Providing practical programmer guidance for working with (or around) those limitations -- such as where it is important to use other complementary techniques such as automatic local variable initialisation.
- Signpost some amount of ongoing research in these areas.

rwatson committed:
First cut at a 'limitations' section to the CHERI C/C++ guide.

rwatson committed:
Minor wording and phrasing improvements / smoothing.

rwatson committed:
Align section title with body text version.

rwatson committed:
Expand on and clarify integer-pointer type confusion.  Fix word-o.

qwattash commented on src/limitations/bounds-imprecision.md:
```diff
@@ -0,0 +1,52 @@
+  sizes and poor alignment within a larger strucure.
+  Compiler warnings about limited precision should be observed.
+  When implementing protecion in memory allocators, guidance provided in this
+  document should be observed to ensure precise spatial safety is achieved.
```
```suggestion
  document should be observed to ensure precise spatial safety is achieved.
  
**Advice to developers**: Sub-allocation patterns should be avoided. If multiple objects need to be allocated, the system allocator (or application-specific allocators) should be used to allocate each disjointed object separately and should guarantee that the allocation is properly aligned and padded for representability. If sub-allocation can not be avoided, care must be taken to ensure that each sub-allocated object is placed at a representable boundary within the main allocation (note that this is platform-specific). 
```

qwattash commented on src/limitations/compile-time-type-uncertainty.md:
```diff
@@ -0,0 +1,105 @@
+capabilities, or whether the destination should or can accept their storage.
+For example, a pointer to a structure that does contain a pointer field could
+be cast to `void *`, losing that information.
+Similarly, a pointer to an array integer types, and no pointer fields, could
```
```suggestion
Similarly, a pointer to an array of integer types, and no pointer fields, could
```

qwattash commented on src/limitations/compile-time-type-uncertainty.md:
```diff
@@ -0,0 +1,105 @@
+be cast to `void *`, losing that information.
+While a manual copy of fields might do so using variables the do (or do not)
+preserve tagged values, `memcpy()` implementations must be *capability
+oblivious*: They copy copy any capabilities present, preserving rather than
```
```suggestion
oblivious*: They copy any capabilities present, preserving rather than
```

qwattash reviewed (commented):
Added some suggestions and typo fixes.

rwatson committed:
Correct typo.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

rwatson committed:
Correct word-o.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

rwatson committed:
Remove dup word.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

rwatson committed:
Fix typo.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

rwatson committed:
Fix word-o.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

rwatson committed:
Fix word-o.

Co-authored-by: Alfredo Mazzinghi <mzz.lrd@gmail.com>

qwattash committed:
limitations: Add discussion on sub-allocation patterns.

Describe a very generic example to explain how this differs from a custom
allocator.
Add developer advice on sub-allocations.

qwattash committed:
limitations: Extend ongoing research to mention exact sub-object bounds.

This is a short summary of early experiments on the CheriBSD kernel.

qwattash commented on src/limitations/bounds-imprecision.md:
```diff
@@ -0,0 +1,52 @@
+  "opportunistic" to being deterministically safe.
+  This involves improvements to compiler analysis, the use of trapping
+  bounds-setting instructions to ensure fail-closedd behavior, and optional
+  support for additional alignment and padding.
```
Added some text in the "ongoing research" paragraph.

qwattash committed:
limitations: Add cross reference to `ptraddr_t`.

Add advice to developers section to suggest the use of `ptraddr_t` in place of
`long` and other integer types to hold a virtual address.

rwatson committed:
Qualify statement on c18n stacks, to match the stack temporal-safety section.

rwatson committed:
Minor wording tweaks.

rwatson committed:
Minor wording tweaks.

rwatson committed:
Minor wording tweaks.

rwatson committed:
Fix typo.  Mention kernel nocap memcpy(), and propose for userlevel.

rwatson commented on src/limitations/compile-time-type-uncertainty.md:
```diff
@@ -0,0 +1,105 @@
+  integer in the same type (perhaps as an opaque argument), or to enable more
+  rich forms of arithmetic on pointers.
+  `long` and `unsigned long` should never be used to hold pointers that must
+  remain deferenceable.
```
Closing out this thread since a patch has been proposed.

davidchisnall commented on src/limitations/bounds-imprecision.md:
```diff
@@ -0,0 +1,78 @@
+   must similarly take into account new alignment and padding requirements.
+
+Where possible, the compiler will emit warnings in situations where sub-object
+bounds cannot be guaranteed to provide precise spatial protection.
```
We also have a static analyser pass that warns for cases where sub-object bounds would not be precisely representable.  This is something we'll be aiming to upstream to mainline LLVM.

qwattash replied:
We should sync to avoid duplicating work, since we have also been discussing work on similar things.

rwatson replied:
I will expand the note on ongoing research to also mention SCI.

rwatson committed:
Add note on ptraddr_t and it ongoing standardisation.

Co-authored-by: Brooks Davis <brooks@one-eyed-alien.net>

davidchisnall reviewed (commented):
Meta comment: One sentence per line makes review much easier for documentation.

rwatson committed:
Fix typo.

Flowdalic commented on src/limitations/unions.md:
```diff
@@ -0,0 +1,40 @@
+mitigating such potential vulnerabilities, but as this does not provide strong
+type safety, nor prevent other types of type confusion, ultimately this is
+mitigation against attack techniques, rather than elimination of
+ulnerability.
```
```suggestion
vulnerability.
```

rwatson committed:
Fix spelling.

Co-authored-by: Florian Schmaus <florian.schmaus@codasip.com>

rwatson committed:
Mention SCI as doing ongoing research around sub-object bounds.

rwatson committed:
Slight rewording.

rwatson committed:
Try to expand a bit the comment on integer-pointer confusion with intptr_t.

rwatson commented on src/limitations/compile-time-type-uncertainty.md:
```diff
@@ -0,0 +1,116 @@
+     used only for pointer types.
+     However, this means that CHERI C/C++ are not able to provide dynamic
+     integer-pointer type-confusion prevention, as the types are inherently
+     ambiguous.
```
I’ve attempted to explore this in a bit more detail in a6ea5530ff7c722af3832ad4061b30d202237106; could you review this change and see if it helps / propose further improvements?

rwatson committed:
Loosen language on sizeof(intptr_t) != sizeof(intmax_t).

rwatson commented on src/limitations/compile-time-type-uncertainty.md:
```diff
@@ -0,0 +1,116 @@
+
+  1. Capability-sized strorage will be allocated, rather than that of the
+     largest integer type, which can be confusing; i.e., `sizeof(intptr_t)`
+     is not the same as `sizeof(intmax_t)`.
```
Now done.

rwatson committed:
Add further examples of uses of sealing, explicitly, to improve robustness.

rwatson commented on src/limitations/integer-pointer-safety.md:
```diff
@@ -0,0 +1,56 @@
+  opportunity for pointer-pointer type confusion, and careful tagging and
+  checking of invariants dynamically.
+
+**Ongoing research**: SRI/Cambridge have been exploring dynamic enforcement
+  of language-level types using the architectural *otype* feature, but have
+  yet to find a satisfactory middle ground between idiomatic type flexibility
+  and strong vulnerability mitigation.
```
I’ve written this up somewhat in 0120944f4ec232e5b23383aa56772a144e71eab5; on the whole, this report has not talked about compartmentalisation to date, and now we are adding some minor references that this is consistent with, but at some point we may well need to think about saying something in a more structured way. Do take a look and feel free to fix the above change further?

rwatson committed:
Improve heading consistency with ToC.

rwatson committed:
Further heading and ToC/outline consistency.
